### ISSUE TITLE
Does `--stdin-filename` actually do anything?

### ISSUE THREAD

**Issue by fosskers** [astral-sh/ruff#17307]
### Question

In configuring `ruff` for use in my editor (Emacs w/ apheleia), I've noticed that specifying the command as follows works without issue:
```
ruff format --stdin-filename foo --line-length=120
```
The file to be saved is piped in through stdin, and it doesn't seem to matter what I set `--stdin-filename` to, it can be anything, and the reformatting still works. If that's the case, what is that flag for? Note that without it, `ruff` ignores stdin and just formats the whole directory.

### Version

0.11.4

**Comment by MichaReiser** [comment]
The filename is also used to infer the file type. E.g. a file ending in `.ipynb` is treated as a notebook and a file ending in `.pyi` is a stub file. Different rules apply to those files.

**Comment by fosskers** [comment]
Is there a world where piping to stdin "just works" without having to specify that flag?

**Comment by MichaReiser** [comment]
I think it should work if you use `ruff format -` which also enables piping

**Comment by fosskers** [comment]
A lot of the classic Unix tools seem to, so there must be a way.

**Comment by Skylion007** [comment]
We can detect if the stdin input to ruff is connected to a TTY with rust's stdlib [now](https://doc.rust-lang.org/beta/std/io/trait.IsTerminal.html):
```rust
use std::io::{self, IsTerminal, Read};

fn main() {
    let stdin = io::stdin();

    if stdin.is_terminal() {
        println!("No piped input detected. Stdin is from a terminal.");
    } else {
        println!("Piped input detected. Reading from stdin...");

        let mut input = String::new();
        stdin.read_to_string(&mut input).unwrap();
        println!("Received input:\n{}", input);
    }
}
```
or via atty crate on older rust versions, I think our

**Comment by sharkdp** [comment]
No, I don't think atty checks are relevant here. They can be used to check whether stdin is attached to a *interactive terminal* or not. But they can't be used to check whether "the user intends to pass something on stdin". The only implicit thing that you could check whether or not stdin is open or closed. But it is extremely rare for stdin to be closed. You would basically have to do something like `ruff … < /dev/null`.

So the only way to potentially support this would be to *explicitly* say that the "no positional argument" call of "ruff check" should read from stdin. But I don't think that is an option. The "no positional argument" call currently has a (arguably) much more useful default, which is to check the current directory.

So I don't think there's anything that can be done here. Users will need to pass `-` explicitly if they are in the (rare) use case where they want to pass something on stdin.

**Comment by sharkdp** [comment]
> Is there a case where stdin is open on a non-interactive session and user doesn't intend to pass input though?

That's what I tried to explain. stdin is almost always open, unless you close it explicitly. And it should be! You can do `cat 1.txt - 3.txt > combined.txt`, and `cat` will read from those two files, but also read from stdin. Or a program could suddenly decide to ask you to enter something in an interactive prompt.

So when you do `ruff check path/to/explicit/file.py`, stdin is open. There's nothing preventing `ruff` from also reading from stdin in that case.

> Or at least we could throw a warning if users have an open stdin in a non-interactive input AND files specified?

No, that would be a terrible UX.

> IE. We should warn maybe if user have files specified in a non-interactive terminal and is piping to std input?

I don't think that would be a good idea. You might run `ruff check` from a script or some other environment where stdin is not a TTY. It shouldn't behave different in those cases.
